### PR TITLE
Feature: InjectiveStaking add new compound_rewards function

### PR DESCRIPTION
**Pull request by defser**
This pull request introduces a new feature to the `InjectiveStaking` module: the `compound_rewards` function. This function enables users to withdraw staking rewards from a validator and automatically re-stake them, compounding the rewards to maximize staking benefits.

---

### 1. New Functionality:
- **`compound_rewards(validator_address: str)`**
  - Automates:
    - Fetching the current balance.
    - Withdrawing staking rewards.
    - Calculating available rewards after fees.
    - Re-delegating the rewards back to the validator.

### 2. Error Handling:
- Introduced specific error messages for scenarios:
  - No rewards are available for compounding.
  - Rewards are insufficient to cover gas fees.
  - General errors with detailed exceptions.

### 3. Optimizations:
- Added a utility function `wait_for_balance_update` to ensure updated balance information after withdrawals.

---

## How to Test
1. Ensure the agent has staked INJ with a validator and accumulated sufficient rewards.
2. Call the `compound_rewards` function with a valid `validator_address`.
3. Validate the following scenarios:
   - Successful reward compounding.
   - Handling of insufficient rewards (reward balance lower than gas fees).
   - Handling of cases where no rewards are available.

---

## Benefits
- Automates reward compounding, reducing manual steps.
- Improves user experience by handling edge cases and errors gracefully.
- Provides a scalable foundation for further staking-related enhancements.

---

## Notes
- Ensure your wallet has sufficient funds to cover gas fees for both withdrawal and delegation transactions.
- This feature relies on accurate balance updates after withdrawal; any delays might require revisiting the `wait_for_balance_update` implementation.

---

## Checklist
- [x] Added `compound_rewards` function.
- [x] Implemented error handling for edge cases.
- [x] Tested for common and edge scenarios.
- [x] Updated documentation for the new functionality.

---

Feel free to suggest improvements or additional scenarios to test!

### CODE REVIEW

File: injective_functions/staking/staking_schema.json [hunk 1]
```diff
@@ -17,6 +17,20 @@
           },
           "required": ["validator_address", "amount"]
         }
+      },
+      {
+        "name": "compound_rewards",
+        "description": "Automatically reinvest your staking rewards with a specific validator to increase your staked amount.",
```
[the remark this answers is not in the document]
Comment (defser) [reply injective_functions/staking/staking_schema.json:23]: This is the current convention for declaring functions within this setup, and unfortunately, it's not possible to modify it at this time.

File: injective_functions/staking/__init__.py [hunk 2]
```diff
@@ -19,3 +20,89 @@ async def stake_tokens(self, validator_address: str, amount: str) -> Dict:
...
+                # Generic error for zero rewards
+                return {"success": False, "error": "No rewards available to compound."}
+
+            # Step 5: Restake the rewards
+            delegate_msg = self.chain_client.composer.MsgDelegate(
+                delegator_address=self.chain_client.address.to_acc_bech32(),
+                validator_address=validator_address,
+                amount=rewards_to_stake / Decimal('1e18'),
+            )
+            delegate_response = await self.chain_client.build_and_broadcast_tx(delegate_msg)
```
[the remark this answers is not in the document]
Comment (defser) [reply injective_functions/staking/__init__.py:70]: This already happens when calling build_and_broadcast_tx:

injective_functions/utils/initializers.py:57

sim_res = await self.client.simulate(sim_tx_raw_bytes)

File: injective_functions/staking/__init__.py [hunk 3]
```diff
@@ -19,3 +20,89 @@ async def stake_tokens(self, validator_address: str, amount: str) -> Dict:
...
+                denom="inj"
+            )
+            initial_balance = Decimal(balance_response.balance.amount)
+
+            # Step 2: Withdraw rewards
+            withdraw_msg = self.chain_client.composer.msg_withdraw_delegator_reward(
+                delegator_address=self.chain_client.address.to_acc_bech32(),
+                validator_address=validator_address,
+            )
+            withdraw_response = await self.chain_client.build_and_broadcast_tx(withdraw_msg)
```
[the remark this answers is not in the document]
Comment (defser) [reply injective_functions/staking/__init__.py:46]: This already happens when calling build_and_broadcast_tx:

injective_functions/utils/initializers.py:57

sim_res = await self.client.simulate(sim_tx_raw_bytes)

**Comment by defser** [comment]
All coderabbitai comments resolved.

File: injective_functions/staking/__init__.py [hunk 4]
```diff
@@ -19,3 +20,89 @@ async def stake_tokens(self, validator_address: str, amount: str) -> Dict:
...
+
+            # Step 3: Wait for the balance to update
+            updated_balance = await self.wait_for_balance_update(old_balance=initial_balance, denom="inj")
+
+            # Step 4: Calculate the withdrawn rewards
+            rewards_to_stake = updated_balance - initial_balance
```
Comment (djosse) [inline injective_functions/staking/__init__.py:52]: Please clean up this logic path to cover each discrete case, something like
```
rewards_to_stake = updated_balance - initial_balance

if rewards_to_stake < 0:
    return {
        "success": False,
        "error": f"Rewards ({rewards_to_stake}) are lower than gas fees, resulting in a negative net reward."
    }

if rewards_to_stake == 0:
    return {"success": False, "error": "No rewards available to compound."}
```

File: injective_functions/staking/__init__.py [hunk 5]
```diff
@@ -19,3 +20,89 @@ async def stake_tokens(self, validator_address: str, amount: str) -> Dict:
...
+                                 f"reward."
+                    }
+                # Generic error for zero rewards
+                return {"success": False, "error": "No rewards available to compound."}
+
+            # Step 5: Restake the rewards
+            delegate_msg = self.chain_client.composer.MsgDelegate(
+                delegator_address=self.chain_client.address.to_acc_bech32(),
+                validator_address=validator_address,
+                amount=rewards_to_stake / Decimal('1e18'),
```
Comment (djosse) [inline injective_functions/staking/__init__.py:66]: I haven't checked other parts of the code here but suggest creating a constant for the 1e18 decimal conversion

File: injective_functions/functions_schemas.json [hunk 6]
```diff
@@ -778,6 +778,20 @@
         ]
       }
     },
+    {
```
Comment (djosse) [inline injective_functions/functions_schemas.json:781]: Why does this schema exist in both function_schema and functions_schema? Actually why are there two separate schema files?

File: injective_functions/staking/staking_schema.json [hunk 7]
```diff
@@ -17,6 +17,20 @@
           },
           "required": ["validator_address", "amount"]
         }
+      },
```
Comment (djosse) [inline injective_functions/staking/staking_schema.json:20]: Can you consolidate the schemas into one location? Looks like this one shows up in 3 different schema locations.

**Review by djosse — COMMENTED** [review]
Please review.

File: injective_functions/staking/__init__.py (hunk 5 above)
Comment (defser) [reply injective_functions/staking/__init__.py:66]: I’ve defined the constant within the InjectiveStaking class for now, as it’s currently only being used in this context. If the need arises to use it more broadly in the future, we can consider moving it to a shared location for better accessibility.

File: injective_functions/staking/staking_schema.json (hunk 7 above)
Comment (defser) [reply injective_functions/staking/staking_schema.json:20]: Agreed, I’ve placed it only inside staking_schema.json.

File: injective_functions/functions_schemas.json (hunk 6 above)
Comment (defser) [reply injective_functions/functions_schemas.json:781]: I’ve placed it only inside staking_schema.json.

**Comment by defser** [comment]
> Please review.

Comments resolved

**Comment by defser** [comment]
@djosse please review

**Comment by defser** [comment]
> @defser can you please rebase with master?

Done
